### PR TITLE
8296246: Update Unicode Data Files to Version 15.1.0

### PR DESCRIPTION
**Pull request by naotoj**
This PR is to incorporate the latest Unicode 15.1, which was released yesterday. Besides the usual character data update, an upgraded implementation of RegEx which reflects the Indic Conjunct Break specified in the latest [Unicode Annex #29 ("Unicode Text Segmentation")](https://unicode.org/reports/tr29/) is included. A corresponding CSR has also been drafted.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8316217](https://bugs.openjdk.org/browse/JDK-8316217) to be approved

### Issues
 * [JDK-8296246](https://bugs.openjdk.org/browse/JDK-8296246): Update Unicode Data Files to Version 15.1.0 (**Enhancement** - P3)
 * [JDK-8316217](https://bugs.openjdk.org/browse/JDK-8316217): Update Unicode Data Files to Version 15.1.0 (**CSR**)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [31e397ce](https://git.openjdk.org/jdk/pull/15728/files/31e397ceac3bd36d64b090961b47c513b401243b)
 * [Joe Wang](https://openjdk.org/census#joehw) (@JoeWang-Java - **Reviewer**)
 * [Steven Loomis](https://openjdk.org/census#srl) (@srl295 - Committer)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [31e397ce](https://git.openjdk.org/jdk/pull/15728/files/31e397ceac3bd36d64b090961b47c513b401243b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15728/head:pull/15728` \
`$ git checkout pull/15728`

Update a local copy of the PR: \
`$ git checkout pull/15728` \
`$ git pull https://git.openjdk.org/jdk.git pull/15728/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15728`

View PR using the GUI difftool: \
`$ git pr show -t 15728`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15728.diff">https://git.openjdk.org/jdk/pull/15728.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15728#issuecomment-1718266502)

### CODE REVIEW

File: make/jdk/src/classes/build/tools/generateextraproperties/GenerateExtraProperties.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+ *     (cp >= 0x2000 && cp <= 0x200A);
+ * </blockquote>
+ *
+ * Arguments to this utility:
+ *    args[0]: Full path string to the template file
+ *    args[1]: Full path string to the properties file
+ *    args[2]: Full path string to the generated .java file
+ *    args[3...]: Names of the property to generate the conditions
+ */
+public class  GenerateExtraProperties {
```
Comment (turbanoff) [inline make/jdk/src/classes/build/tools/generateextraproperties/GenerateExtraProperties.java:65]:
```suggestion
public class GenerateExtraProperties {
```

File: make/jdk/src/classes/build/tools/generateextraproperties/GenerateExtraProperties.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                    StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    static String rangeToString(Range r) {
+        if (r.start == r.last) {
+            return (" ".repeat(12) + "cp == 0x" + toHexString(r.start));
+        } else  if (r.start == r.last - 1) {
```
Comment (turbanoff) [inline make/jdk/src/classes/build/tools/generateextraproperties/GenerateExtraProperties.java:118]:
```suggestion
        } else if (r.start == r.last - 1) {
```

File: make/modules/java.base/gensrc/GensrcRegex.gmk [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#
+# Rules to create java files under
+# $(SUPPORT_OUTPUTDIR)/gensrc/java.base/jdk/internal/util/regex/
+#
+
+GENSRC_INDICCONJUNCTBREAK := $(SUPPORT_OUTPUTDIR)/gensrc/java.base/jdk/internal/util/regex/IndicConjunctBreak.java
+
+INDICCONJUNCTBREAKTEMP = $(MODULE_SRC)/share/classes/jdk/internal/util/regex/IndicConjunctBreak.java.template
+INDICCONJUNCTBREAKPROPS = $(MODULE_SRC)/share/data/unicodedata/DerivedCoreProperties.txt
+INDICCONJUNCTBREAKPARAMS = InCB\=Linker InCB\=Extend InCB\=Consonant
```
Comment (erikj79) [inline make/modules/java.base/gensrc/GensrcRegex.gmk:35]: Please use `:=` for assignment unless you explicitly need lazy evaluation for these variables. I can't see any reason to need that here.

I don't think you need to escape `=`. What happens if you don't?

File: make/modules/java.base/gensrc/GensrcRegex.gmk [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+$(GENSRC_INDICCONJUNCTBREAK): $(BUILD_TOOLS_JDK) $(INDICCONJUNCTBREAKTEMP) $(INDICCONJUNCTBREAKPROPS)
+	$(call LogInfo, Generating $@)
+	$(call MakeTargetDir)
+	$(TOOL_GENERATEEXTRAPROPERTIES) \
+		$(INDICCONJUNCTBREAKTEMP) \
+		$(INDICCONJUNCTBREAKPROPS) \
+		$(GENSRC_INDICCONJUNCTBREAK) \
+		$(INDICCONJUNCTBREAKPARAMS)
```
Comment (erikj79) [inline make/modules/java.base/gensrc/GensrcRegex.gmk:44]: Please use 4 spaces after the initial tab for continuation indentation in recipes (see point 2 and 6 https://openjdk.org/groups/build/doc/code-conventions.html).

File: make/modules/java.base/gensrc/GensrcRegex.gmk (hunk 3 above)
Comment (naotoj) [reply make/modules/java.base/gensrc/GensrcRegex.gmk]: You are correct on both. Fixed.

File: make/modules/java.base/gensrc/GensrcRegex.gmk (hunk 4 above)
Comment (naotoj) [reply make/modules/java.base/gensrc/GensrcRegex.gmk]: I wasn't aware of the convention (initial tab, then spaces for further indentation). Thanks.

**Review by erikj79 — APPROVED** [review]
Build changes look good.

File: make/jdk/src/classes/build/tools/generateextraproperties/GenerateExtraProperties.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Parses extra properties files of UCD, and replaces the placeholders in
+ * the given template source file with the generated conditions, then emit
```
Comment (JoeWang-Java) [inline make/jdk/src/classes/build/tools/generateextraproperties/GenerateExtraProperties.java:41]: s/emit/emits?

File: make/modules/java.base/gensrc/GensrcRegex.gmk [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#
+# Rules to create java files under
+# $(SUPPORT_OUTPUTDIR)/gensrc/java.base/jdk/internal/util/regex/
+#
+
+GENSRC_INDICCONJUNCTBREAK := $(SUPPORT_OUTPUTDIR)/gensrc/java.base/jdk/internal/util/regex/IndicConjunctBreak.java
+
+INDICCONJUNCTBREAKTEMP := $(MODULE_SRC)/share/classes/jdk/internal/util/regex/IndicConjunctBreak.java.template
+INDICCONJUNCTBREAKPROPS := $(MODULE_SRC)/share/data/unicodedata/DerivedCoreProperties.txt
+INDICCONJUNCTBREAKPARAMS := InCB=Linker InCB=Extend InCB=Consonant
```
Comment (JoeWang-Java) [inline make/modules/java.base/gensrc/GensrcRegex.gmk:35]: DerivedCoreProperties.txt is very large. Would it be helpful to reference the section, that is "Derived Property: Indic_Conjunct_Break", in this file or the template?

**Review by JoeWang-Java — APPROVED** [review]
LGTM. A couple of minor suggestions.

File: make/jdk/src/classes/build/tools/generateextraproperties/GenerateExtraProperties.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+ * the produced .java files. For example, if the properties file has:
```
Comment (srl295) [inline make/jdk/src/classes/build/tools/generateextraproperties/GenerateExtraProperties.java:42]:
```suggestion
 * Parses extra properties files of UCD, and replaces the placeholders in
 * the given template source file with the generated conditions, then emits
 * .java files. For example, if the properties file has:
```

File: src/java.base/share/legal/unicode.md [hunk 8]
```diff
@@ -1,103 +1,104 @@
...
-Data Files and Software.
-
-NOTICE TO USER: Carefully read the following legal agreement.
-BY DOWNLOADING, INSTALLING, COPYING OR OTHERWISE USING UNICODE INC.'S
-DATA FILES ("DATA FILES"), AND/OR SOFTWARE ("SOFTWARE"),
-YOU UNEQUIVOCALLY ACCEPT, AND AGREE TO BE BOUND BY, ALL OF THE
-TERMS AND CONDITIONS OF THIS AGREEMENT.
-IF YOU DO NOT AGREE, DO NOT DOWNLOAD, INSTALL, COPY, DISTRIBUTE OR USE
-THE DATA FILES OR SOFTWARE.
+UNICODE LICENSE V3
```
Comment (srl295) [inline src/java.base/share/legal/unicode.md:6]: very  minor update

File: src/java.base/share/legal/unicode.md [hunk 9]
```diff
@@ -1,103 +1,104 @@
...
+
+Except as contained in this notice, the name of a copyright holder shall
+not be used in advertising or otherwise to promote the sale, use or other
+dealings in these Data Files or Software without prior written
+authorization of the copyright holder.
+
+
+------------------
+
+Unicode® Copyright and Terms of Use
```
Comment (srl295) [inline src/java.base/share/legal/unicode.md:49]: You shouldn't need to pull in all of this — the entire license file is at https://www.unicode.org/license.txt 

It looks like you are pulling in all of https://www.unicode.org/copyright.html

You will notice that the license text no longer points you to https://www.unicode.org/copyright.html

**Review by srl295 — APPROVED** [review]

File: src/java.base/share/legal/unicode.md (hunk 9 above)
Comment (srl295) [reply src/java.base/share/legal/unicode.md:49]: look at https://www.unicode.org/Public/UCD/latest/ for example, it points you only to license.txt.

**Review by RogerRiggs — APPROVED** [review]
Looks good;  Its not easy to map all the code changes to spec changes.

**Review by srl295 — APPROVED** [review]

File: make/jdk/src/classes/build/tools/generateextraproperties/GenerateExtraProperties.java (hunk 5 above)
Comment (naotoj) [reply make/jdk/src/classes/build/tools/generateextraproperties/GenerateExtraProperties.java]: Thanks. Fixed it along with @srl295 's suggestion

File: make/modules/java.base/gensrc/GensrcRegex.gmk (hunk 6 above)
Comment (naotoj) [reply make/modules/java.base/gensrc/GensrcRegex.gmk:35]: Added explanation in `IndicConjunctBreak.java.template`

File: src/java.base/share/legal/unicode.md (hunk 9 above)
Comment (naotoj) [reply src/java.base/share/legal/unicode.md:49]: We are checking on it and will update it if necessary.

File: src/java.base/share/classes/jdk/internal/util/regex/Grapheme.java [hunk 10]
```diff
@@ -35,9 +35,10 @@ public final class Grapheme {
      * <p>
      * See Unicode Standard Annex #29 Unicode Text Segmentation for the specification
      * for the extended grapheme cluster boundary rules. The following implementation
-     * is based on the annex for Unicode version 15.0.
-     * (http://www.unicode.org/reports/tr29/tr29-40.html)
+     * is based on the annex for Unicode version 15.1.
+     * (http://www.unicode.org/reports/tr29/tr29-43.html)
      *
+     * @spec http://www.unicode.org/reports/tr29/tr29-43.html
```
[the remark this answers is not in the document]
Comment (naotoj) [reply src/java.base/share/classes/jdk/internal/util/regex/Grapheme.java]: Thanks, removed.

File: src/java.base/share/legal/unicode.md (hunk 9 above)
Comment (srl295) [reply src/java.base/share/legal/unicode.md:49]: it was unnecessary to mention copyright.html or terms of use in the source license previously

**Review by JoeWang-Java — APPROVED** [review]

**Pull request closed without merging**